### ISSUE TITLE
TestCopySimple fails

### ISSUE THREAD

**Issue by elboulangero** [tonistiigi/fsutil#26]
```
# go get github.com/tonistiigi/fsutil
# go test github.com/tonistiigi/fsutil -v -run ''


=== RUN   TestCopySimple
--- FAIL: TestCopySimple (0.01s)
	assertions.go:238: 
			Error Trace:	receive_test.go:80
			Error:      	Not equal: 
			            	expected: "sha256:2c6454be696ff984b27888e4757cfb5eb65c1a8f0d49c017596e3feefaa0c84e"
			            	actual  : "sha256:99b6ef96ee0572b5b3a4eb28f00b715d820bfd73836e59cc1565e241f4d1bb2f"
	assertions.go:238: 
			Error Trace:	receive_test.go:84
			Error:      	Not equal: 
			            	expected: "sha256:2fcd2d8fb162dbdfb2ecfe93724dc0f4e683ebff2d9f15485136958ef99ef8b7"
			            	actual  : "sha256:dd2529f7749ba45ea55de3b2e10086d6494cc45a94e57650c2882a6a14b4ff32"
	assertions.go:238: 
			Error Trace:	receive_test.go:88
			Error:      	Not equal: 
			            	expected: "sha256:51086b730b712a6c3a11651083e4eb00a095b3ef60174fb8934542e8cebcecca"
			            	actual  : "sha256:eca07e8f2d09bd574ea2496312e6de1685ef15b8e6a49a534ed9e722bcac8adc"
	assertions.go:238: 
			Error Trace:	receive_test.go:137
			Error:      	Not equal: 
			            	expected: "sha256:51086b730b712a6c3a11651083e4eb00a095b3ef60174fb8934542e8cebcecca"
			            	actual  : "sha256:eca07e8f2d09bd574ea2496312e6de1685ef15b8e6a49a534ed9e722bcac8adc"
	assertions.go:238: 
			Error Trace:	receive_test.go:141
			Error:      	Not equal: 
			            	expected: "sha256:f2b015030e257b821d76681d822e23af06de1621e3b1a9168fbc9d22e83ed345"
			            	actual  : "sha256:cd14a931fc2e123ded338093f2864b173eecdee578bba6ec24d0724272326c3a"
```

Test with both go 1.7 and go 1.10, in case it matters.

**Comment by thaJeztah** [comment]
Are you running the tests on a Mac? I see the same when running the tests on my Mac, but not when running inside a Linux container

**Comment by thaJeztah** [comment]
Adding some rudimentary debugging;

```patch
diff --git a/receive_test.go b/receive_test.go
index 78bb2f4..e3c73dd 100644
--- a/receive_test.go
+++ b/receive_test.go
@@ -3,6 +3,8 @@ package fsutil
 import (
        "bytes"
        "crypto/sha256"
+       "encoding/json"
+       "fmt"
        "hash"
        io "io"
        "io/ioutil"
@@ -320,6 +322,10 @@ func simpleSHA256Hasher(s *Stat) (hash.Hash, error) {
        h := sha256.New()
        ss := *s
        ss.ModTime = 0
+
+       jsonString, _ := json.Marshal(ss)
+       fmt.Println(string(jsonString))
+
        dt, err := ss.Marshal()
        if err != nil {
                return nil, err
```

Shows that one difference is the uid/gid (expected, because the files are created by the current user, which is "root" inside the container);


### On macOS:

```
=== RUN   TestCopySimple
{"path":"foo","mode":420,"uid":503,"gid":20,"size":5}
{"path":"foo2","mode":420,"uid":503,"gid":20,"size":4}
{"path":"zzz","mode":2147484096,"uid":503,"gid":20,"size":136}
{"path":"zzz/aa","mode":420,"uid":503,"gid":20,"size":5}
{"path":"zzz/bb","mode":2147484096,"uid":503,"gid":20,"size":102}
{"path":"zzz/bb/cc","mode":2147484096,"uid":503,"gid":20,"size":102}
{"path":"zzz/bb/cc/dd","mode":134218221,"uid":503,"gid":20,"size":6,"linkname":"../../"}
{"path":"zzz.aa","mode":420,"uid":503,"gid":20}
{"path":"zzz/bb/cc/foo","mode":384,"uid":503,"gid":20,"size":5}
--- FAIL: TestCopySimple (0.01s)
	Error Trace:	receive_test.go:125
	Error:      	Not equal: 
	            	expected: "sha256:99b6ef96ee0572b5b3a4eb28f00b715d820bfd73836e59cc1565e241f4d1bb2f"
	            	actual: "sha256:a32f799e80725555be3b46135d01745bcef933e98781e2183a135cf5fe2d68fe"
	Error Trace:	receive_test.go:129
	Error:      	Not equal: 
	            	expected: "sha256:dd2529f7749ba45ea55de3b2e10086d6494cc45a94e57650c2882a6a14b4ff32"
	            	actual: "sha256:a258ef51018680cceac4867ab34764ba26ef6c75b9dc60a2d44676d62240c5fd"
	Error Trace:	receive_test.go:133
	Error:      	Not equal: 
	            	expected: "sha256:eca07e8f2d09bd574ea2496312e6de1685ef15b8e6a49a534ed9e722bcac8adc"
	            	actual: "sha256:b9b0f4540c62aafeb5af599ad58ecfe569cbb03ecddcc67a7a180bbaf07a10c6"
	Error Trace:	receive_test.go:183
	Error:      	Not equal: 
	            	expected: "sha256:eca07e8f2d09bd574ea2496312e6de1685ef15b8e6a49a534ed9e722bcac8adc"
	            	actual: "sha256:b9b0f4540c62aafeb5af599ad58ecfe569cbb03ecddcc67a7a180bbaf07a10c6"
	Error Trace:	receive_test.go:187
	Error:      	Not equal: 
	            	expected: "sha256:cd14a931fc2e123ded338093f2864b173eecdee578bba6ec24d0724272326c3a"
	            	actual: "sha256:a825c23fc3ff7fd58f4b2a0290a73da074a96d627a88146b4abb8397247f2397"
FAIL

Process finished with exit code 1
```


### In a Linux container:

```bash
docker run -it --rm --workdir=/go/src/github.com/tonistiigi/fsutil/ -v $(pwd):/go/src/github.com/tonistiigi/fsutil golang
go get -t -v ./...
go test github.com/tonistiigi/fsutil -v -run ''
```

```
=== RUN   TestCopySimple
{"path":"foo","mode":420,"size":5}
{"path":"foo2","mode":420,"size":4}
{"path":"zzz","mode":2147484096,"size":4096}
{"path":"zzz/aa","mode":420,"size":5}
{"path":"zzz/bb","mode":2147484096,"size":4096}
{"path":"zzz/bb/cc","mode":2147484096,"size":4096}
{"path":"zzz/bb/cc/dd","mode":134218239,"size":6,"linkname":"../../"}
{"path":"zzz.aa","mode":420}
{"path":"zzz/bb/cc/foo","mode":384,"size":5}
```

So, one difference is ownership, which would be expected. When resetting the `uid` and `gid`:

```patch
diff --git a/receive_test.go b/receive_test.go
index 78bb2f4..325e7eb 100644
--- a/receive_test.go
+++ b/receive_test.go
@@ -320,6 +320,8 @@ func simpleSHA256Hasher(s *Stat) (hash.Hash, error) {
        h := sha256.New()
        ss := *s
        ss.ModTime = 0
+       ss.Uid = 0
+       ss.Gid = 0
        dt, err := ss.Marshal()
        if err != nil {
                return nil, err
```

Comes closer, but there's still a difference;

```patch
diff --git a/linux.txt b/macos2.txt
index 6e1b35a..0387dad 100644
--- a/linux.txt
+++ b/macos2.txt
@@ -1,10 +1,10 @@
 {"path":"foo","mode":420,"size":5}
 {"path":"foo2","mode":420,"size":4}
-{"path":"zzz","mode":2147484096,"size":4096}
+{"path":"zzz","mode":2147484096,"size":136}
 {"path":"zzz/aa","mode":420,"size":5}
-{"path":"zzz/bb","mode":2147484096,"size":4096}
-{"path":"zzz/bb/cc","mode":2147484096,"size":4096}
-{"path":"zzz/bb/cc/dd","mode":134218239,"size":6,"linkname":"../../"}
+{"path":"zzz/bb","mode":2147484096,"size":102}
+{"path":"zzz/bb/cc","mode":2147484096,"size":102}
+{"path":"zzz/bb/cc/dd","mode":134218221,"size":6,"linkname":"../../"}
 {"path":"zzz.aa","mode":420}
 {"path":"zzz/bb/cc/foo","mode":384,"size":5}
```

Size difference could possibly be related to block-size?

Not sure about the `mode` difference:

 - Linux: `134218239` (octal: `1000000777`)
 - macOS: `134218221` (octal: `1000000755`)

**Comment by thaJeztah** [comment]
fixed through https://github.com/tonistiigi/fsutil/issues/37 - thanks for reporting!

**Comment by elboulangero** [comment]
@thaJeztah 

> Are you running the tests on a Mac? I see the same when running the tests on my Mac, but not when running inside a Linux container

Nope, I'm running these tests on my laptop, running Debian unstable.

I can confirm that it's all fixed with #38, thanks!
